hal/watchdog_api.h

```diff
@@ -0,0 +1,171 @@
+   * the independent watchdog. In this mode a restriction is placed on the
+   * time period in which the watchdog can be refreshed/kicked. If the
+   * watchdog is kicked too soon after it has last been refreshed the system
+   * will be reset. The period of time in which the reset will be triggered is
+   * defined by window_ms. This value is false by default.
+   */
+  bool enable_window;
+  /**
+   * Specifies the time window for the watchdog window in milliseconds. If the
+   * watchdog is configured to run in windowed mode kicking the watchdog less
+   * than this many milliseconds after it has last been kicked will trigger a
+   * system reset. This value must be less than timeout_ms.
+   */
+  uint32_t window_ms;
+  /**
+   * Configures the watchdog behaviour while the system is in sleep mode. When
+   * this flag is enabled the watchdog timer runs normally while the system is
+   * in sleep mode, when disabled the watchdog is paused during this time.
+   */
+  bool enable_sleep;
+} watchdog_config_t;
+
+
+typedef enum {
+  WATCHDOG_STATUS_OK,
+  WATCHDOG_STATUS_NOT_SUPPORTED,
+  WATCHDOG_STATUS_INVALID_ARGUMENT
+} watchdog_status_t;
+
+#ifdef __cplusplus
+    extern "C" {
+#endif
+
+/** Initialise and start a watchdog timer with the given configuration.
+ *
+ * If the watchdog timer is configured and started successfully this
+ * function will return WATCHDOG_STATUS_OK.
+ *
+ * If the enable_window is set but windowed mode is not supported by the
+ * platform the function will return WATCHDOG_STATUS_NOT_SUPPORTED.
+ *
+ * If the timeout specified is outside the range supported by the platform,
+ * or if the window period is greater than the timeout period it will return
+ * WATCHDOG_STATUS_INVALID_ARGUMENT.
+ *
+ * @param[in]  config   Configuration settings for the watchdog timer
+ *
+ * @return WATCHDOG_STATUS_OK if the watchdog is configured correctly and
+ *         has started. Otherwise a status indicating the fault.
+ */
+watchdog_status_t hal_watchdog_init(const watchdog_config_t *config);
+
+/** Refreshes the watchdog timer.
+ *
+ * This function should be called periodically before the watchdog times out.
+ * Otherwise, the system is reset.
+ *
+ * If using the windowed operation mode this function must be called within the
+ * configured timeout window. If the function is called before the window is
+ * reached the system is reset.
+ *
+ * If a watchdog is not currently running this function does nothing
+ */
+void hal_watchdog_kick(void);
+
+/** Stops the watchdog timer
+ *
+ * Calling this function will attempt to disable any currently running watchdog
+ * timers if supported by the current platform
+ *
+ * @return Returns WATCHDOG_STATUS_OK if the watchdog timer was succesfully
+ *         stopped, or if the timer was never started. Returns
+ *         WATCHDOG_STATUS_NOT_SUPPORTED if the watchdog cannot be disabled on
+ *         the current platform.
+ */
+watchdog_status_t hal_watchdog_stop(void);
```

Comment:
Do we want to support watchdogs which can't be turned off? This could be particularly problematic for devices with a bootloader which enables the watchdog. Any application running with this kind of a bootloader will be unable to turn the watchdog off.

Also, for these type of devices is the watchdog interval fixed as well? If so this should probably be exposed in the watchdog API.

Reply:
It's not uncommon for platforms to not support disabling the watchdog after it's started, or at least it's not documented how. I think it's best to inform the user that it can't be stopped somehow, when they're configuring it, or just when they try to stop it instead of just not supporting that platform.

The intervals for the watchdog are configurable on all platforms I checked, but the maxiumum refresh values vary wildly between the platforms, 30 odd seconds for STM vs a few years on K64F.